ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java

```diff
@@ -197,4 +225,313 @@ public void updateAccounts(Collection<Account> accounts, int batchSize) throws S
+   */
+  public void updateAccountsAndContainers(List<AccountUtils.AccountUpdateInfo> accountsInfo, int batchSize)
+      throws SQLException {
+    Connection conn = dataAccessor.getDatabaseConnection(true);
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      PreparedStatement insertAccountStatement = dataAccessor.getPreparedStatement(insertAccountsSql, true);
+      PreparedStatement updateAccountStatement = dataAccessor.getPreparedStatement(updateAccountsSql, true);
+      PreparedStatement insertContainerStatement = dataAccessor.getPreparedStatement(insertContainersSql, true);
+      PreparedStatement updateContainerStatement = dataAccessor.getPreparedStatement(updateContainersSql, true);
+
+      conn.setAutoCommit(false);
+      int batchCount = 0, insertAccountCount = 0, updateAccountCount = 0, insertContainerCount = 0,
+          updateContainerCount = 0;
+      for (int i = 0; i < accountsInfo.size(); i++) {
+        Account account = accountsInfo.get(i).getAccount();
+        boolean isAccountModified = accountsInfo.get(i).isModified();
+        List<Container> addedContainers = accountsInfo.get(i).getAddedContainers();
+        List<Container> updatedContainers = accountsInfo.get(i).getModifiedContainers();
+
+        if (isAccountModified) {
+          if (account.getSnapshotVersion() == 0) {
+            insertAccountStatement.setString(1, AccountCollectionSerde.accountToJsonNoContainers(account).toString());
+            insertAccountStatement.setInt(2, account.getSnapshotVersion());
+            insertAccountStatement.addBatch();
+            ++insertAccountCount;
+          } else {
+            updateAccountStatement.setString(1, AccountCollectionSerde.accountToJsonNoContainers(account).toString());
+            updateAccountStatement.setInt(2, account.getSnapshotVersion());
+            updateAccountStatement.setInt(3, account.getId());
+            updateAccountStatement.addBatch();
+            ++updateAccountCount;
+          }
+          ++batchCount;
+        }
+
+        for (Container container : addedContainers) {
+          insertContainerStatement.setInt(1, account.getId());
+          insertContainerStatement.setString(2, container.toJson().toString());
+          insertContainerStatement.setInt(3, container.getSnapshotVersion());
+          insertContainerStatement.addBatch();
+          ++insertContainerCount;
+          if (++batchCount >= batchSize) {
+            executeBatchUpdates(insertAccountCount > 0 ? insertAccountStatement : null,
+                updateAccountCount > 0 ? updateAccountStatement : null,
+                insertContainerCount > 0 ? insertContainerStatement : null,
+                updateContainerCount > 0 ? updateContainerStatement : null);
+            conn.commit();
+            batchCount = 0;
+            insertAccountCount = 0;
+            updateAccountCount = 0;
+            insertContainerCount = 0;
+            updateContainerCount = 0;
+          }
```

Comment:
This logic is repeated a few times.  How about if we create an inner class like AccountUpdateBatch encapsulating the statements and counters.  And it can have a method like maybeExecuteBatch() with this block of code.